### ISSUE TITLE
Overload matchers

### ISSUE THREAD

**Issue by maxmeyer** [rspec/rspec-expectations#797]
Hi @myronmarston,

I'm using `capybara` for some projects + `aruba`. Capybara has a nice `have_content`-matcher. For `aruba` I implemented a similar thing but named it differently to prevent a name-clash. While refactoring `aruba` a lot I asked myself what would be the best to support `have_content`-matchers for different things.

Just for `aruba` I could put together something like cucumber/aruba#248 but "not" for other projects. Are there any recommendations for that? Would it make sense to put together a PR, which makes such a thing possible?

```ruby
RSpec::Matchers.define :have_content do |expected|
  select_if do |actual|
    actual.is_a? File # or /pattern/ === actual
  end

  match do |actual|
    path = absolute_path(actual)

    next false unless File.file? path

    content = File.read(path).chomp
    next expected === content if expected.is_a? Regexp

    content == expected.chomp
  end

  diffable

  failure_message do |actual|
    next format("expected that file \"%s\" contains:\n%s", actual, expected) if expected.is_a? Regexp

    format("expected that file \"%s\" contains exactly:\n%s", actual, expected)
  end

  failure_message_when_negated do |actual|
    next format("expected that file \"%s\" does not contain:\n%s", actual, expected) if expected.is_a? Regexp

    format("expected that file \"%s\" does not contains exactly:\n%s", actual, expected)
  end
end
```

Ping @mattwynne @jarl-dk. :smile:

**Comment by myronmarston** [comment]
I don't understand your question.  Can you rephrase it?

**Comment by JonRowe** [comment]
He's asking us to add conditional selection of matcher implementations based on the block provided to `select_if`, I don't think this is something we should do on a general basis. The same effect could be done by overriding the method yourself and returning the relevant matcher, but the performance cost would probably be too great to have as a core feature IMO

**Comment by myronmarston** [comment]
> He's asking us to add conditional selection of matcher implementations based on the block provided to `select_if`

Thanks.  I somehow didn't notice the `select_if` block in the example at all...

> I don't think this is something we should do on a general basis.

I agree.  This can be accomplished already with Ruby.  Any DSL support provided by RSpec would just be "sugar" that wouldn't provide significant benefit.  I don't think this is a common enough use case to warrant RSpec providing a DSL for it.

**Comment by maxmeyer** [comment]
Ok. Thanks for the feedback.
